renku/core/models/provenance/provenance_graph.py

```diff
@@ -0,0 +1,321 @@
+    def activities(self):
+        """Return a map from order to activity."""
+        return {a.order: a for a in self._activities}
+
+    def add(self, node: Union[Activity, ActivityCollection]):
+        """Add an Activity/ActivityCollection to the graph."""
+        assert self._loaded
+
+        activity_collection = node if isinstance(node, ActivityCollection) else ActivityCollection(activities=[node])
+        # filename = str(Path(filename).name)
+        # path = self.provenance_files_path / filename
+        # activity_collection.to_json(path=path)
+
+        for activity in activity_collection._activities:
+            assert not any([a for a in self._activities if a.id_ == activity.id_]), f"Identifier exists {activity.id_}"
+            activity.order = self._order
+            self._order += 1
+            self._activities.append(activity)
+
+    @classmethod
+    def from_json(cls, path, lazy=False):
+        """Return an instance from a YAML file."""
+        if Path(path).exists():
+            if not lazy:
+                with open(path) as file_:
+                    data = json.load(file_)
+                    self = cls.from_jsonld(data=data)
+                    self._activities.sort(key=lambda e: e.order)
+            else:
+                self = ProvenanceGraph(activities=[])
+                self._loaded = False
+        else:
+            self = ProvenanceGraph(activities=[])
+            self._loaded = True
+
+        self._path = path
+
+        return self
+
+    @classmethod
+    def from_jsonld(cls, data):
+        """Create an instance from JSON-LD data."""
+        if isinstance(data, cls):
+            return data
+        elif not isinstance(data, list):
+            raise ValueError(data)
+
+        self = ProvenanceGraphSchema(flattened=True).load(data)
+        self._loaded = True
+
+        return self
+
+    def to_jsonld(self):
+        """Create JSON-LD."""
+        return ProvenanceGraphSchema(flattened=True).dump(self)
+
+    def to_json(self, path=None):
+        """Write an instance to file."""
+        path = path or self._path
+        data = self.to_jsonld()
+        with open(path, "w", encoding="utf-8") as file_:
+            json.dump(data, file_, ensure_ascii=False, sort_keys=True, indent=2)
+
+    @property
+    def rdf_graph(self):
+        """Create an RDFLib ConjunctiveGraph."""
+        self._create_rdf_graph()
+        return self._graph
+
+    def _create_rdf_graph(self):
+        # import json
+        # import pyld
+        # from rdflib.plugin import Parser, register
+        # output = pyld.jsonld.expand([a.as_jsonld() for a in self._activities])
+        # data = json.dumps(output, indent=2)
+        # register("json-ld", Parser, "rdflib_jsonld.parser", "JsonLDParser")
+        if self._graph:
+            return
+
+        self._graph = ConjunctiveGraph().parse(location=str(self._path), format="json-ld")
+
+        self._graph.bind("prov", "http://www.w3.org/ns/prov#")
```

Comment:
a small feedback: it'd be great if the bindings of schemas could be done in one place (see these lines and https://github.com/SwissDataScienceCenter/renku-python/pull/1545/files#diff-2c509c92e0e0a8e64101568aa3de495e48895669678cae50d1888703b43f6d83R141). these are redundant lines...

in addition having the class ctor (init) function have an optional list of tuples of additional binding pairs would allow something like this:

```
def bindings(self, graph):
    # default bindings
    graph.bind("foaf", "http://xmlns.com/foaf/0.1/")
    ...
    # custom bindings
    for i in self.custom_bindings:
        graph.bind(i[0], i[1])
```

otherwise queries could be a bit redundant if one uses schemas that are not in the default binding list

Reply:
Thanks Viktor for the comment and code!

One of the functions was removed and now the binding is done in one place. I also used your code for annotations and custom binding (although I've used a dict instead of a list): https://github.com/SwissDataScienceCenter/renku-python/pull/1545/commits/1c6be49e068086c80b464d1fddf9c2490b5c921e.